### PR TITLE
Cache config to improve latencies

### PR DESCRIPTION
**Pull request by ecktom**
## Related issue
Fixes https://github.com/ory/oathkeeper/issues/346

Blocked by https://github.com/ory/viper/pull/7

## Proposed changes

Did a first draft here. Also had to hash the entire ENV in order to allow overrides which made things like 30-40% slower. ;) Any idea for that?

Guess we could tweak things even more eg. 
- pre-allocate + copy slices instead of append
- simple bit shift instead of binary.LittleEndian 

which however would make readability worse.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: driver/configuration/provider_viper.go [hunk 1]
```diff
@@ -197,7 +204,53 @@ func (v *ViperProvider) pipelineIsEnabled(prefix, id string) bool {
...
 func (v *ViperProvider) PipelineConfig(prefix, id string, override json.RawMessage, dest interface{}) error {
+	hash, err := v.hashPipelineConfig(prefix, id, override)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	c, ok := v.configCachge[hash]
+	if ok {
+		if dest != nil {
+			marshalled, err := json.Marshal(c)
```
Comment (aeneasr) [inline driver/configuration/provider_viper.go]: Let's safe one round trip of encoding here - see line https://github.com/ory/oathkeeper/pull/348/files#diff-2449a6ea083767b149400840c05f41bdR308

I thought it might be a good idea to use glob, so I wrote a small benchmark. Unfortunately, glob seems to underperform when having to deal with interfaces (**so let's stick to json**), here are the results:

```
go test -bench . -benchmem
goos: darwin
goarch: amd64
pkg: github.com/ory/oathkeeper/docs
BenchmarkGobJSON/gob-12                     7848            151588 ns/op           55343 B/op       1234 allocs/op
BenchmarkGobJSON/json-12                   20253             57361 ns/op           32480 B/op        489 allocs/op
```

```go
package docs

import (
	"bytes"
	"encoding/gob"
	"encoding/json"
	"io/ioutil"
	"testing"

	"github.com/ghodss/yaml"
	"github.com/stretchr/testify/require"
)

func BenchmarkGobJSON(b *testing.B) {
	gob.Register(map[string]interface{}{})
	gob.Register([]interface{}{})

	c, err := ioutil.ReadFile(".oathkeeper.yaml")
	require.NoError(b, err)

	conf, err := yaml.YAMLToJSON(c)
	require.NoError(b, err)

	var o map[string]interface{}
	require.NoError(b, json.NewDecoder(bytes.NewBuffer(conf)).Decode(&o))

	var gb bytes.Buffer
	require.NoError(b, gob.NewEncoder(&gb).Encode(&o))

	b.Run("gob", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			b.StopTimer()
			m := make(map[string]interface{})
			buf := bytes.NewBuffer(gb.Bytes())
			b.StartTimer()

			require.NoError(b, gob.NewDecoder(buf).Decode(&m))
		}
	})

	b.Run("json", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			b.StopTimer()
			m := make(map[string]interface{})
			buf := bytes.NewBuffer(conf)
			b.StartTimer()

			require.NoError(b, json.NewDecoder(buf).Decode(&m))
		}
	})
}
```

File: driver/configuration/provider_viper.go [hunk 2]
```diff
@@ -305,7 +299,13 @@ func (v *ViperProvider) PipelineConfig(prefix, id string, override json.RawMessa
 		return errors.WithStack(result.Errors())
 	}
 
-	v.configCachge[hash] = config
+	marshalled, err = json.Marshal(config)
```
Comment (aeneasr) [inline driver/configuration/provider_viper.go]: `marshalled` should already be the proper type and payload (see https://github.com/ory/oathkeeper/pull/348/files/ede19d9b718fa4d91191ae9d138e89546305440c..8231b9aa87680589b8c42c25a46c5188acc446ca#diff-2449a6ea083767b149400840c05f41bdR296) so I don't think we need to encode it again!

File: driver/configuration/provider_viper_public_test.go [hunk 3]
```diff
@@ -48,11 +48,9 @@ func TestPipelineConfig(t *testing.T) {
 		require.NoError(t, p.PipelineConfig("authenticators", "oauth2_introspection", nil, &res))
 		assert.JSONEq(t, `{"introspection_request_headers":{},"introspection_url":"https://override/path","pre_authorization":{"client_id":"some_id","client_secret":"some_secret","enabled":true,"scope":["foo","bar"],"token_url":"https://my-website.com/oauth2/token"},"required_scope":[],"scope_strategy":"exact","target_audience":[],"trusted_issuers":[]}`, string(res), "%s", res)
 
+		// Cleanup
 		require.NoError(t, os.Setenv("AUTHENTICATORS_OAUTH2_INTROSPECTION_CONFIG_INTROSPECTION_URL", ""))
 
-		require.NoError(t, p.PipelineConfig("authenticators", "oauth2_introspection", nil, &res))
```
Comment (aeneasr) [inline driver/configuration/provider_viper_public_test.go]: Why were these removed?

File: driver/configuration/provider_viper_public_test.go [hunk 4]
```diff
@@ -93,6 +91,38 @@ func TestPipelineConfig(t *testing.T) {
...
 }
 
+/*
+go test ./... -v -bench=. -run BenchmarkPipelineConfig -benchtime=10s
+
+v0.35.1
+594	  20119202 ns/op
+
+v0.35.2
+3048037	      3908 ns/op
```
Comment (aeneasr) [inline driver/configuration/provider_viper_public_test.go]: Wow, that perf improvement!

File: driver/configuration/provider_viper_public_test.go (hunk 3 above)
Comment (ecktom) [reply driver/configuration/provider_viper_public_test.go]: This was basically the root cause of my intention to cache the environment ;) But as you've correctly pointed out there will be no changes to the env for the running process... So I've simply adjusted the test to reflect this.

File: driver/configuration/provider_viper.go (hunk 2 above)
Comment (ecktom) [reply driver/configuration/provider_viper.go]: yeah :D quiet obviously... Removed the unnecessary marshall...

File: driver/configuration/provider_viper_public_test.go (hunk 4 above)
Comment (ecktom) [reply driver/configuration/provider_viper_public_test.go]: Indeed :D

File: driver/configuration/provider_viper_public_test.go (hunk 3 above)
Comment (aeneasr) [reply driver/configuration/provider_viper_public_test.go]: Ohh, I see - that totally makes sense :)

**Comment by aeneasr** [comment]
Awesome, thank you so much for your hard work!
